### ISSUE TITLE
Middleware-like usage?

### ISSUE THREAD

**Issue by 00ff0000red** [yhirose/cpp-httplib#813]
Is there any way that I could set up a handler that just intercepts all requests, sends some headers or something, and returns, moving on to the handler for that specific URL?

Or will I have to set up the function calls myself?

**Comment by yhirose** [comment]
@00ff0000red, thanks for the feedback, but I don't quite understand what you are asking for. Could you explain it more clearly? Thanks!

**Comment by PixlRainbow** [comment]
Something like `set_logger` but not const and runs before the path-specific handler instead of after.

**Comment by yhirose** [comment]
@PixlRainbow, thank you for helping me understand the issue.

@00ff0000red, I experimentally made the following methods in `middleware` branch:
https://github.com/yhirose/cpp-httplib/pull/816/files

```cpp
void set_pre_routing_handler(Handler handler);
void set_post_routing_handler(Handler handler);
```

Do these satisfy your needs?

**Comment by 00ff0000red** [comment]
> Something like `set_logger` but not const and runs before the path-specific handler instead of after.

Yes, that describes it well! My partner and I were using the library for HTTP, but eventually, we realized that we were going to need a server that provides middleware methods. Actually, I don't even know what it's called, or if "middleware" is the correct word for it.

My partner suggested that the lib has loggers, but it didn't fit our use case.

My use case:: generally I like to prevent/redirect external requests before doing any "business logic" e.g. to, say, prevent attackers from messing with my sites.
WIth some server libraries, this would manifest itself as a handler that checks whether it is an outside request or not, if it is, redirect them, otherwise move on to the appropriate handler for that URL.

Ex, Node express.js:
```js
server.get("*", (request, response, next) => {
    if ( ...logic here... ) {
        response.status(403).render("...file...");
    } else {
        next();
    }
});

server.get("/", (request, response) => {
    request.sendFile("../frontend/index.xhtml");
});
```

So, how about making `set_pre_routing_handler` return a `bool` that determines whether or not to continue the response or move on?

**Comment by Yrds** [comment]
So instead `pre_routing_handler -> regular handler(Response to client happen here) -> post_routing_handler(called normally)`

You want be possible to `pre_routing_handler(Response to client can happen here) -> regular handler(Can be never called) -> post_routing_handler(called normally)` by some logic in `pre_routing_handler`

Something like this?

**Comment by 00ff0000red** [comment]
> So instead `pre_routing_handler -> regular handler(Response to client happen here) -> post_routing_handler(called normally)`
> 
> You want be possible to `pre_routing_handler(Response to client can happen here) -> regular handler(Can be never called) -> post_routing_handler(called normally)` by some logic in `pre_routing_handler`
> 
> Something like this?

Oh, I meant just to break the entire chain, ex:
```c++
pre_routing_handler(Response to client can happen here) -> regular handler(Can be never called) -> post_routing_handler(Can be never called)
```

What do you think would be a sensible API though? I guess someone might want control over both situations?

**Comment by 19wintersp** [comment]
Could it be made an option to just call `res.end()` (or some other method) at any point to break the chain?

**Comment by Yrds** [comment]
> > So instead `pre_routing_handler -> regular handler(Response to client happen here) -> post_routing_handler(called normally)`
> > You want be possible to `pre_routing_handler(Response to client can happen here) -> regular handler(Can be never called) -> post_routing_handler(called normally)` by some logic in `pre_routing_handler`
> > Something like this?
> 
> Oh, I meant to just break the entire chain, ex:
> 
> ```c++
> pre_routing_handler(Response to client can happen here) -> regular handler(Can be never called) -> post_routing_handler(Can be never called)
> ```
> 
> What do you think would be a sensible API though? I guess someone might want control over both situations?

```
#include "httplib.h"

class Middleware {
  public:
    Middleware(httplib::Server::Handler next){
      this->next = next;
    }

    httplib::Server::Handler handler() {
      auto handlerFunc = [this](const httplib::Request &req, httplib::Response &res) {
        if(req.params.find("validParam") == req.params.end()){
          const std::string content = " \
          <html> \
            <body> \
              Hi Foo, insert validParam to see bar \
            </body> \
          </html> \
          ";
          res.set_content(content, "text/html");
        }
        else {
          next(req, res);
        }
      };

      return handlerFunc;
  }
  private:
    httplib::Server::Handler next;

};

int main(){

  httplib::Server svr;

  auto regularHandler = [](const httplib::Request &req, httplib::Response &res) {
      const std::string content = " \
      <html> \
        <body> \
          Hi Bar, remove validParam to see foo \
        </body> \
      </html> \
      ";
      res.set_content(content, "text/html");
  };


  svr.Get("/test", (Middleware(regularHandler).handler()));


  svr.listen("0.0.0.0", 8080);

  return 0;
}
```

In this example if you send `http://localhost:8080/test` will call a generic handler from Middleware class, but if you send `http://localhost:8080/test?validParam` then will call the `regularHandler` normally. 

Of course this middleware class is very simple and can be extended but i made it quickly just do demonstrate the behaviour.


Something like this but for a global entire `Server` (or predefined paths)?

**Comment by 00ff0000red** [comment]
@Yrds' Example is good, but do you want that in the library itself, or should someone just make a wrapper around the library that does that? Seems like it could be a major change, and it would require a good deal of tests to make sure everything is working as intended.

**Comment by Yrds** [comment]
> > So instead `pre_routing_handler -> regular handler(Response to client happen here) -> post_routing_handler(called normally)`
> > 
> > You want be possible to `pre_routing_handler(Response to client can happen here) -> regular handler(Can be never called) -> post_routing_handler(called normally)` by some logic in `pre_routing_handler`
> > 
> > Something like this?
> 
> Oh, I meant to just break the entire chain, ex:
> ```c++
> pre_routing_handler(Response to client can happen here) -> regular handler(Can be never called) -> post_routing_handler(Can be never called)
> ```
> 
> What do you think would be a sensible API though? I guess someone might want control over both situations?

   Yes i meant something through the library API, i am not sure how `Server::routing` works but it's a good idea investigate if is a good idea for implement this behavior in this part of code or add the capability of interrupt the request in `pre_routing_handler`.

**Comment by 00ff0000red** [comment]
Also, not sure about explicitly calling `next`, since each call would consume stack space, unless the compiler optimized into return calls (and platform support). Boolean returns might be better.

**Comment by yhirose** [comment]
@00ff0000red, I thought you just need opportunities to modify request/response objects before processing a request or sending a response. So the pre/post_request handles are not meant to handle chaining of request path matching. 

I now understand what you are trying to do in your project. I'll ponder over it to find an effective way to implement the 'routing middleware' feature without breaking the current API.

**Comment by 00ff0000red** [comment]
> I thought you just need opportunities to modify request/response objects before processing a request or sending a response. So the pre/post_request handles are not meant to handle chaining of request path matching.

That does cover half of what I do, like I had said, just setting headers to all incoming requests, but the other part of it, chaining, isn't covered. So, they are still helpful,

If you decide that you don't want to try this, for whatever reason, the pre and post handlers would still prove to be very helpful.

**Comment by yhirose** [comment]
@00ff0000red, could you check the latest httplib.h in the middleware branch?

I added `bool` return value to `pret_request_handler`.
https://github.com/yhirose/cpp-httplib/blob/2099c475f9e6d92a241de13aa14599ef55372a4e/httplib.h#L600

So you could do like this:
```cpp
svr.set_pre_routing_handler([] (const Request &req, Response &res) {
  if (/* logic here...*/) {
    res.body = "handled in the pre rounting handler...";
    res.status = 403;
    return true;
  }

  // let the router handle the request
  return false;
});
```

This is obviously not as powerful as the express.js router middleware. But I think it can satisfy all your needs and it doesn't require any existing interface change. (`post_routing_handler` stays same, because it shouldn't affect the routing control flow.)

Could you test with it? If it works for you, I'll merge it and bump up the version of cpp-httplib.
Thanks!

**Comment by 00ff0000red** [comment]
it seems like it'll be all that I need, it fires before every request, it takes a single function, therefore preventing my current duplicate code, and can prevent moving forward to the normal handler. Although I was thinking of the boolean return value in the opposite way, the return value would decide whether or not the function chain should continue being called, whereas yours means whether it was handled in the handler, just a different way of thinking about it, but it's fine.

I'll try out the branch tomorrow and report back.

**Comment by yhirose** [comment]
@00ff0000red, thanks for the confirmation.

> As for the C++17 thing, I thought that you wanted to remove code and simplify maintenance? Wouldn't it be easier to just freeze the current API, and create a new branch dedicated to C++17 where could use any newer features from the stdlib to make the API simpler, and the header smaller?

I was initially thinking of removing the current API completely when introducing the new API, but it seems like a number of other projects now depend on this library. So I would rather freeze the existing API instead of removing it, so that the change won't affect those projects and give users time to migrate to the new API. But I'll make any additional enhancements only on the new API. It won't happen anytime soon anyway though, since I don't have enough time to work on it for now.

As soon as I hear a positive report from you, I'll merge it to the main branch. Thanks.

**Comment by 00ff0000red** [comment]
@yhirose Works just fine!

**Comment by 00ff0000red** [comment]
Update: doesn't work at all, server starts then stops instantly, I'll try to repro.

**Comment by yhirose** [comment]
@00ff0000red, any update?

**Comment by 00ff0000red** [comment]
I tried to reproduce the next day in a simple, isolated project, couldn't do it.
It is possible that there was something else going on in my code, I'll try to pin it down tomorrow, I hadn't edited the server's code at all today.
Sorry, for the late response.

**Comment by 00ff0000red** [comment]
Okay, today I copied my code to begin breaking it apart to find the source, but it worked right away, so I don't know what was wrong.
The same source code gave different results, maybe it was my build process?

I guess this is safe to merge, the pre_handler seemed to work fine (so far).

**Comment by yhirose** [comment]
Implemented by #816

**Comment by heysurfer** [comment]
```c++
auto AuthCheck = [=] (const httplib::Request & req, httplib::Response & res)
    {
        if (!auth()) {
            res.status = 401;
            return httplib::Server::HandlerResponse::Handled;
        }
        return httplib::Server::HandlerResponse::Unhandled;
    };
svr.set_pre_routing_handler(AuthCheck);
```
